File: luiz/ba3d.py
#! /usr/bin/env python

from ba3c import overlap
from ba3a import composition


def de_bruijn(k, text):
    kmers = list(composition(k - 1, text))

    dbg = {}

    for kmer, neighbor in overlap(kmers):
        if kmer not in dbg:
            dbg[kmer] = set()

        dbg[kmer].add(neighbor)

    return dbg


assert de_bruijn(4, "AAGATTCTCTAC") == {
    "AAG": {"AGA"},
    "AGA": {"GAT"},
    "ATT": {"TTC"},
    "CTA": {"TAC"},
    "CTC": {"TCT"},
    "GAT": {"ATT"},
    "TCT": {"CTA", "CTC"},
    "TTC": {"TCT"},
}


if __name__ == "__main__":
    with open("data/rosalind_ba3d.txt", "r") as dataset:
        k = int(dataset.readline().strip())
        text = dataset.readline().strip()

    dbg = de_bruijn(k, text)

    for kmer, neighbors in dbg.items():
        if len(neighbors) == 1:
            print(kmer, "->", neighbors.pop())
        else:
            neighbors = list(neighbors)
            print(kmer, "->", neighbors[0], end="")
            for neighbor in neighbors[1:]:
                print(",{}".format(neighbor))

File: luiz/src/bin/ba3d.rs
use std::collections::HashMap;
use std::env;
use std::error::Error;
use std::fs;

use bioinformatics_algorithms::{composition, overlap};

pub fn de_bruijn(k: usize, text: &str) -> HashMap<String, Vec<String>> {
    let kmers: Vec<&str> = composition(k - 1, text).collect();

    let mut dbg = HashMap::default();

    for (kmer, neighbor) in overlap(&kmers) {
        let neighbors = dbg.entry(kmer).or_insert(Vec::with_capacity(4));
        neighbors.push(neighbor);
    }

    dbg
}

fn main() -> Result<(), Box<dyn Error>> {
    let input: String = env::args()
        .nth(1)
        .unwrap_or("data/rosalind_ba3d.txt".into());
    let data = fs::read_to_string(input)?;

    let mut lines = data.lines();
    let k = lines.next().unwrap().parse()?;
    let text: String = lines.next().unwrap().into();

    for (node, neighbors) in de_bruijn(k, &text) {
        if neighbors.len() == 1 {
            println!("{} -> {}", node, neighbors[0]);
        } else {
            print!("{} -> {}", node, neighbors[0]);

            for neighbor in &neighbors[1..] {
                print!(",{}", neighbor);
            }
            println!();
        }
    }

    Ok(())
}
